compiler/rustc_lint/src/default_could_be_derived.rs

```diff
@@ -0,0 +1,422 @@
+    /// }
+    ///
+    /// #[deny(default_could_be_derived)]
+    /// impl Default for A {
+    ///     fn default() -> A {
+    ///         A {
+    ///             b: None,
+    ///             c: Some(0),
+    ///         }
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// `#[derive(Default)]` uses the `Default` impl for every field of your
+    /// type. If your manual `Default` impl either invokes `Default::default()`
+    /// or uses the same value that that associated function produces, then it
+    /// is better to use the derive to avoid the different `Default` impls from
+    /// diverging over time.
+    ///
+    /// This lint also triggers on cases where there the type has no fields,
+    /// so the derive for `Default` for a struct is trivial, and for an enum
+    /// variant with no fields, which can be annotated with `#[default]`.
+    pub DEFAULT_COULD_BE_DERIVED,
+    Warn,
+    "detect `Default` impl that could be derived"
+}
+
+declare_lint! {
+    /// The `manual_default_for_type_with_default_fields` lint checks for
+    /// manual `impl` blocks of the `Default` trait of types with default
+    /// field values.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// #![feature(default_field_values)]
+    /// struct Foo {
+    ///     x: i32 = 101,
+    ///     y: NonDefault,
+    /// }
+    ///
+    /// struct NonDefault;
+    ///
+    /// #[deny(default_overrides_default_fields)]
+    /// impl Default for Foo {
+    ///     fn default() -> Foo {
+    ///         Foo { x: 100, y: NonDefault }
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Manually writing a `Default` implementation for a type that has
+    /// default field values runs the risk of diverging behavior between
+    /// `Type { .. }` and `<Type as Default>::default()`, which would be a
+    /// foot-gun for users of that type that would expect these to be
+    /// equivalent. If `Default` can't be derived due to some fields not
+    /// having a `Default` implementation, we encourage the use of `..` for
+    /// the fields that do have a default field value.
+    pub DEFAULT_OVERRIDES_DEFAULT_FIELDS,
+    Warn,
+    "detect `Default` impl that should use the type's default field values"
+}
+
+declare_lint_pass!(DefaultCouldBeDerived => [DEFAULT_COULD_BE_DERIVED, DEFAULT_OVERRIDES_DEFAULT_FIELDS]);
+
+impl<'tcx> LateLintPass<'tcx> for DefaultCouldBeDerived {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx hir::Item<'tcx>) {
+        if !cx.tcx.features().default_field_values() {
+            // FIXME: this logic can be extended to check for types without default field values
+            // doing the same as `clippy::derivable_impls`, with slightly more information thanks
+            // to the added "default equivalence" table.
+            return;
+        }
```

Comment:
@compiler-errors the nightly gate

Reply:
It should also be gated in the lint def so users cannot name them with `allow`/`deny` without a feature gate.